Fix Unchecked Arithmetic in Diem-Core

terrytzh opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

Making money move more freely for more people!

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

## Related PRs

None

## If targeting a release branch, please fill the below out as well

Fix unchecked arithmetic to prevent overflow.

JoshLind commented on secure/storage/vault/src/lib.rs:
```diff
@@ -369,7 +371,13 @@ impl Client {
+                .checked_sub(MAX_NUM_KEY_VERSIONS)
+                .and_then(|n| n.checked_add(1))
+                .ok_or_else(|| {
+                    Error::OverflowError(String::from("trim_key_versions::min_available_version"))
```
nit: you could also use `into` to do the string conversion for you. It's shorter: https://doc.rust-lang.org/std/convert/trait.Into.html, e.g.,:

`Error::OverflowError("trim_key_versions::min_available_version".into())`

JoshLind commented on secure/net/src/lib.rs:
```diff
@@ -499,12 +501,14 @@ impl NetworkStream {
-            unwritten = &data[total_written..];
+            total_written = total_written
+                .checked_add(written as u8)
+                .ok_or_else(|| Error::OverflowError(String::from("write_all::total_written")))?;
```
nit: you could use `into`. See my comment below.

JoshLind commented on secure/storage/vault/src/lib.rs:
```diff
@@ -50,6 +50,8 @@ pub enum Error {
     #[error("Synthetic error returned: {0}")]
     SyntheticError(String),
+    #[error("Overflow error: {0}")]
+    OverflowError(String),
```
We normally try to keep these in alphabetical order. Could you move it up?

JoshLind commented on secure/net/src/lib.rs:
```diff
@@ -144,6 +144,8 @@ pub enum Error {
     #[error("Remote stream cleanly closed")]
     RemoteStreamClosed,
+    #[error("Overflow error: {0}")]
+    OverflowError(String),
```
We normally try to keep these in alphabetical order. Could you move it up?

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -126,23 +140,25 @@ mod test {
 
     #[test]
-    fn test_hex_encode() {
+    fn test_hex_encode() -> Result<(), Error> {
```
For tests, you don't need to return the result. Instead, we often just use `unwrap` or `expect` to ensure that an error is not thrown. If one is, the test will panic and fail. For example you could just do:

`hex_encode(&src, &mut actual).unwrap()`

It also keeps the tests cleaner/simpler.

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
+fn byte2hex(byte: u8) -> Result<(u8, u8), Error> {
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
+        .ok_or_else(|| Error::OverflowError(String::from("byte2hex::byte")))?
```
nit: you could use `into`. See my comment below.

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
+            == src
+                .len()
+                .checked_mul(2)
+                .ok_or_else(|| Error::OverflowError(String::from("hex_encode::src")))?
```
nit: you could use `into`. See my comment below.

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
-fn hex_encode(src: &[u8], dst: &mut [u8]) {
-    debug_checked_precondition!(dst.len() == 2 * src.len());
+fn hex_encode(src: &[u8], dst: &mut [u8]) -> Result<(), Error> {
+    debug_checked_precondition!(
```
The formatting/line-wrapping here looks interesting.. Have you run the auto formatter?

Something like `cargo x check` should work (it's an internal tool that wraps the rust auto formatters). I'll also run canary on the PR to see if the tests pass with these changes 😄

JoshLind reviewed (changes requested):
Thanks, Terry! Looks great. 😄

A couple comments and nits to address, but nothing major. Let's also run the canary process which should run all the land blocking tests and format checks to make sure everything passes.

/canary

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/901585673)

terrytzh commented:
Hi Joshua, just fixed the above mentioned issues and changed the function signatures in short-hex-str back to original as well as switched to using expect in error handling in short-hex-str specifically (don't wanna break codes that use this lib!). Would you help me check if any additional changes are needed & rerun the tests in canary?

bmwill commented on common/short-hex-str/src/lib.rs:
```diff
@@ -16,8 +16,10 @@ use thiserror::Error;
 #[derive(Error, Debug)]
-#[error("Input bytes are too short")]
-pub struct InputTooShortError;
+pub enum Error {
+    #[error("Input bytes are too short")]
+    InputTooShortError,
+}
```
Can you revert this change since you no longer have another error variant?

terrytzh replied:
Sure, reverting it rn

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -82,15 +82,24 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
+        .expect("Integer Overflow at byte2hex")
```
Aah.. I think you misunderstood my last comment here. We indeed want `byte2hex` to return a `Result<(u8, u8), Error>` as you had it before, but in the unit tests, we don't need to return the result when we call this function. We simply just want to call `unwrap()` or `expect()` in the unit test. Does that make sense?

The problem with calling the `expect()` here is that non-test code that relies on this function will panic, which isn't great. So, all you need to do is bring back the old logic here and then for the unit tests that call this function, `unwrap()` there. Does that make sense?

The same goes for `hex_encode` below

JoshLind replied:
Once you've done that you can call `/canary` on this PR and see how it goes!

terrytzh replied:
The two functions originally return u8 tuple and nothing, and the previous changes serve as a temporary error handling measure to line up with error handling in other lib files; changing their return type to Result type at this point might cause all codes that call these two functions to break. I'll commit the other two files for now

terrytzh commented:
/canary

bors-libra commented:
@terrytzh: :key: Insufficient privileges: Not Collaborator

JoshLind commented:
LGTM! Let's canary to see if there's any issues with tests/Libya, and if not, I'll approve!

JoshLind commented:
/canary

JoshLind commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/915830421)

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -16,8 +16,12 @@ use thiserror::Error;
-pub struct InputTooShortError;
+pub enum Error {
+    #[error("Input bytes are too short")]
+    InputTooShortError(),
```
nit: this doesn't need empty brackets (as it takes no arguments), just defining the type should be enough, e.g.:

`InputTooShortError,`

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -27,18 +31,18 @@ impl ShortHexStr {
     ///
     /// Returns `Err(InputTooShortError)` if the input byte slice length is less
```
nit: could you update the comment to say "Returns Err(Error::InputTooShortError)"

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -27,18 +31,18 @@ impl ShortHexStr {
             Ok(Self(dest_bytes))
         } else {
-            Err(InputTooShortError)
+            Err(Error::InputTooShortError())
```
nit: this should become `Err(Error::InputTooShortError)`

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -81,22 +85,32 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
-fn hex_encode(src: &[u8], dst: &mut [u8]) {
-    debug_checked_precondition!(dst.len() == 2 * src.len());
+fn hex_encode(src: &[u8], dst: &mut [u8]) -> Result<(), Error> {
+    debug_checked_precondition!(
```
nit: you could tidy this up by creating a temporary variable, e.g.:

```
    let expected_length = src
        .len()
        .checked_mul(2)
        .ok_or_else(|| Error::OverflowError("hex_encode::src".into()))?;
    debug_checked_precondition!(dst.len() == expected_length);
```

JoshLind commented on secure/net/src/lib.rs:
```diff
@@ -499,12 +501,14 @@ impl NetworkStream {
     fn write_all(&mut self, data: &[u8]) -> Result<(), Error> {
         let mut unwritten = data;
-        let mut total_written = 0;
+        let mut total_written: u8 = 0;
```
I think the problem you're facing is that `u8` isn't large enough. This function is used to write data on a network stream, so it's possible you're hitting `u8::MAX` and this function is failing. Change this to `u64` -- I think that should fix the error here.

JoshLind reviewed (commented):
A few comments -- I think the failures you're seeing might be in relation to a genuine overflow as a result of the fact you limited the number of written bytes to `u8`. See my comments below 😄 

Also, let's merge all your commits into a single commit to keep a cleaner history. All of the commits relate to the same unit of work.

JoshLind commented:
/canary

bors-libra commented:
:sunny: Canary successful

JoshLind commented:
/canary

bors-libra commented:
:sunny: Canary successful

JoshLind reviewed (approved):
Thanks @terrytzh! Looks good! 😄 

I'll approve and land! 🚢

JoshLind commented on common/short-hex-str/src/lib.rs:
```diff
@@ -81,22 +85,30 @@ const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
-    let hi = HEX_CHARS_LOWER[((byte >> 4) & 0x0f) as usize];
+fn byte2hex(byte: u8) -> Result<(u8, u8), Error> {
+    let hi = HEX_CHARS_LOWER[(byte
+        .checked_shr(4)
```
@terrytzh, for what it's worth, @bmwill pointed out that this overflow shouldn't ever be possible given the nature of `u8` and shift right. 

However, until we have some way of notifying the linter/arithmetic checker that we've manually reviewed this overflow and it's okay as-is (I'm not aware of a fine-grained way of notifying the arithmetic checker), let's leave it as is.

JoshLind replied:
Thanks @phlip9!

> surprisingly, the compiler is actually able to (almost) completely remove both the checked arithmetic and error paths (since they're impossible) and my local micro benches only have a ~12% regression here.

What regression are you referring to here? Compiler non-optimizations vs. optimizations, or something else?

Thanks for following up with the revert: https://github.com/diem/diem/pull/8509. Will trust you that these paths are safe 😄

phlip9 replied:
I have an old set of micro benchmarks sitting around for when I wrote short-hex-str. they were never published (they're a bit messy). I just added a new case with the error handling and checked arithmetic and compared the benches with optimizations etc. 😅

JoshLind reviewed (commented):
/land

bors-libra commented:
:exclamation: Land has been canceled due to this PR's base ref being changed. Please issue another Land command if you want to requeue this PR.

JoshLind commented:
/land
